### ISSUE TITLE
[macOS] How to hand over the mouse events to the windows behind the transparent areas

### ISSUE THREAD

**Issue by CaineLew** [libsdl-org/SDL#7617]
I can now create shaped windows, but the transparent areas of the windows can still trigger mouse events. I hope that the transparent areas of the windows cannot trigger mouse events, but instead let the windows behind the transparent areas trigger mouse events. What should I do?

**Comment by slouken** [comment]
There is some more discussion about this in https://github.com/libsdl-org/SDL/issues/8454, and it looks like there is some code that might help in https://stackoverflow.com/questions/65286495/make-a-win32-window-transparent-to-messages-click-touch-keyboard

Please let us know if you figure out the solution here?

**Comment by CaineLew** [comment]
The code for the Windows platform is not suitable for me as I work on the macOS platform. I believe that when creating a window with the SDL_WINDOW_TRANSPARENT flag, the window should possess that feature. Alternatively, adding a window flag could also be a solution.

**Comment by slouken** [comment]
There is an example of how to do this here:
https://cocoawithlove.com/2008/12/drawing-custom-window-on-mac-os-x.html

**Comment by kanjitalk755** [comment]
I'm using `NSWindow.ignoresMouseEvents`.

```objc
#include <SDL3/SDL.h>
#include <AppKit/AppKit.h>

static NSWindow *get_nswindow(SDL_Window *window) {
	SDL_PropertiesID props = SDL_GetWindowProperties(window);
	return (__bridge NSWindow *)SDL_GetProperty(props, "SDL.window.cocoa.window", NULL);
}

static int get_title_height(SDL_Window *window) {
	return -(int)[get_nswindow(window) contentRectForFrameRect:NSZeroRect].size.height;
}

static void set_mouse_ignore(SDL_Window *window, BOOL flag) {
	get_nswindow(window).ignoresMouseEvents = flag;
}

int main(int argc, char *argv[]) {
	if (argc != 2) exit(1);
	if (SDL_Init(SDL_INIT_VIDEO)) exit(1);
	atexit(SDL_Quit);
	SDL_Surface *surface = SDL_LoadBMP(argv[1]);
	if (!surface || surface->format->format != SDL_PIXELFORMAT_ARGB8888) exit(1);
	SDL_Window *window = SDL_CreateWindow("alpha test", surface->w, surface->h, SDL_WINDOW_TRANSPARENT);
	if (!window) exit(1);
	SDL_Renderer *renderer = SDL_CreateRenderer(window, NULL, 0);
	if (!renderer) exit(1);
	SDL_Texture *texture = SDL_CreateTextureFromSurface(renderer, surface);
	if (!texture) exit(1);
	SDL_SetRenderDrawColor(renderer, 0, 0, 0, 0);
	SDL_RenderClear(renderer);
	SDL_RenderTexture(renderer, texture, NULL, NULL);
	SDL_RenderPresent(renderer);
//
	const int TITLE_HEIGHT = get_title_height(window);
	for (int quit = 0; !quit;) {
		SDL_Event e;
		while (SDL_PollEvent(&e)) {
			switch (e.type) {
				case SDL_EVENT_QUIT:
					quit = 1;
					break;
			}
		}
//
		int x, y;
		SDL_GetWindowPosition(window, &x, &y);
		float mx, my;
		SDL_GetGlobalMouseState(&mx, &my);
		x = (int)mx - x;
		y = (int)my - y;
		BOOL flag = FALSE;
		if (x >= 0 && x < surface->w) {
			if (y >= 0 && y < surface->h) {
				uint32_t *p = (uint32_t *)surface->pixels + (surface->pitch >> 2) * y + x;
				flag = !(*p & 0xff000000); // alpha is 0
			}
			else flag = y < -TITLE_HEIGHT;
		}
		set_mouse_ignore(window, flag);
	}
	return 0;
}
```

**Comment by CaineLew** [comment]
> There is an example of how to do this here: https://cocoawithlove.com/2008/12/drawing-custom-window-on-mac-os-x.html

Thank you for your response. I resolved the issue using @kanjitalk755's code, but I still hope to see this functionality encapsulated within SDL.

**Comment by kanjitalk755** [comment]
@CaineLew Could you please modify the title of this issue to make it easier for people who are having the same problem to search for it, for example, as below?

[macOS] How to hand over the mouse events to the windows behind the transparent areas

**Comment by CaineLew** [comment]
@CaineLew Sure, no problem!

**Comment by slouken** [comment]
I've implemented this workaround inside SDL, thanks!

**Comment by kanjitalk755** [comment]
`SDL_SetWindowShape` does not seem to be able to detect transparent areas of a surface if they change dynamically.

If this idea is correct, are there any plans to support it?

**Comment by slouken** [comment]
I've clarified that if your surface changes you should call SDL_SetWindowShape() again in 416803b03.

**Comment by kanjitalk755** [comment]
I tested `SDL_SetWindowShape` on macOS with two images included in the SDL source tree and found a problem.

```C++
#include <SDL3/SDL.h>
#include <string>
#include <vector>

static SDL_HitTestResult SDLCALL ShapeHitTest(SDL_Window *window, const SDL_Point *area, void *userdata)
{
	SDL_Surface *shape = (SDL_Surface *)userdata;
	Uint8 r, g, b, a;

	if (SDL_ReadSurfacePixel(shape, area->x, area->y, &r, &g, &b, &a) == 0) {
		if (a != SDL_ALPHA_TRANSPARENT) {
			/* We'll just make everything draggable */
			return SDL_HITTEST_DRAGGABLE;
		}
	}
	return SDL_HITTEST_NORMAL;
}

int main(int argc, char *argv[]) {
	atexit(SDL_Quit);
	std::vector<std::string> images;
	for (int i = 1; i < argc; i++) {
		images.push_back(argv[i]);
	}
	if (images.empty()) exit(1);
	SDL_Surface *shape = SDL_LoadBMP(images[0].c_str());
	if (!shape) exit(1);
	SDL_Window *window = SDL_CreateWindow("", shape->w, shape->h, SDL_WINDOW_TRANSPARENT | SDL_WINDOW_BORDERLESS);
	if (!window) exit(1);
	SDL_Renderer *renderer = SDL_CreateRenderer(window, NULL, SDL_RENDERER_PRESENTVSYNC);
	if (!renderer) exit(1);
	//
	int index = 0;
	auto update_shape = [&]{
		shape = SDL_LoadBMP(images[index].c_str());
		if (!shape) exit(1);
		if (SDL_SetWindowShape(window, shape) < 0) exit(1);
		if (SDL_SetWindowHitTest(window, ShapeHitTest, shape) < 0) exit(1);
		if (++index >= images.size()) index = 0;
	};
	update_shape();
	for (int quit = 0; !quit;) {
		SDL_Event e;
		while (SDL_PollEvent(&e)) {
			switch (e.type) {
				case SDL_EVENT_KEY_DOWN:
					update_shape();
					break;
				case SDL_EVENT_QUIT:
					quit = 1;
					break;
			}
		}
		SDL_SetRenderDrawColor(renderer, 255, 255, 255, SDL_ALPHA_OPAQUE);
		SDL_RenderClear(renderer);
		SDL_RenderPresent(renderer);
	}
	return 0;
}
```

Build the above source and run it as follows.

```
./shape_test SDL/test/audiofile.bmp SDL/test/soundboard.bmp
```

1) You can drag the white area of `audiofile`.
2) Press any key to switch to `soundboard`.
3) The first click may allow you to drag on the transparent area of `soundboard`.

**Comment by slouken** [comment]
> I tested `SDL_SetWindowShape` on macOS with two images included in the SDL source tree and found a problem.

Fixed in f4d97bdb1, thanks!

**Comment by kanjitalk755** [comment]
Unfortunately, the fix was incomplete.

1. Click on the transparent area of `audiofile`.
1. Return focus to the test program with `command`+`tab`.
1. Press any key to switch to `soundboard`.
1. The first click, unable to drag opaque area of `soundboard`.

**Comment by slouken** [comment]
Nice catch, fixed in 5e6ae85b6, thanks!
